Add tooltip to show the control's id [finishes #101719298]

kozzi opened the pull request:
https://www.pivotaltracker.com/n/projects/1256654/stories/101719298

![tooltips](https://cloud.githubusercontent.com/assets/1885340/13307200/70fa9302-db1e-11e5-8881-9e06faac7712.gif)

Wanted to get some eyes on this before I keep going. The original intention was that this tooltip only showed up in code mode. However after discussing this with @poorvasingal, we decided it felt like a bug that the id tooltip only showed up in code mode and not in design mode. So showing the tooltip in both modes.

TODO:
- [x] Add tests

islemaster commented:
This is looking great, @kozzi!  I'll add a couple thoughts inline on possible extractions and tests.

islemaster commented:
Thanks for the gif preview, by the way.

islemaster commented on apps/src/applab/CrosshairOverlay.js:
```diff
@@ -140,3 +180,23 @@ CrosshairOverlay.prototype.getCoordinateText = function () {
+  var elementIdText = this.props_.mouseoverApplabControlId;
+  if (elementIdText.length > ELEMENT_ID_TEXT_MAX_CHAR) {
+    elementIdText = elementIdText.substr(0, ELEMENT_ID_TEXT_MAX_CHAR) + "...";
+  }
```
This is mostly perfect so I'm going to get picky :grin:

I'd like to see this ellipsify method use available space better - right now, adding an ellipsis to the end of the string will cause the final string to be longer than your specified max length.  Here's the current behavior:

```
(9)  ABCDEFGHI    => ABCDEFGHI     (9)
(10) ABCDEFGHIJ   => ABCDEFGHIJ    (10)
(11) ABCDEFGHIJK  => ABCDEFGHIJ... (13)
(12) ABCDEFGHIJKL => ABCDEFGHIJ... (13)
```

I'd like to see the output stay within our specified length, but not ellipsify if the word fits.  Something like this:

```
(9)  ABCDEFGHI    => ABCDEFGHI  (9)
(10) ABCDEFGHIJ   => ABCDEFGHIJ (10)
(11) ABCDEFGHIJK  => ABCDEFG... (10)
(12) ABCDEFGHIJKL => ABCDEFG... (10)
```

Then, assuming the 13-character ellipsified strings were fitting okay in the UI before, you can safely increase your max char constant to 13 and show longer IDs without ellipses. I _think_ this would be closer to the behavior of the CSS3 `text-overflow: ellipsis;` property.

I know this is a trivial bit of logic, but it would be a perfect place for [functional core testing](https://www.destroyallsoftware.com/talks/boundaries).  Here's what you do:
1. Extract a static helper `CrosshairOverlay.ellipsifyText = function (inputText, maxLength) {...}`
2. Write your unit tests against that static method.  Since it's a [pure function](https://en.wikipedia.org/wiki/Pure_function) testing it won't require any special set-up, fakes, etc.

kozzi replied:
Fixed and updated the new max length. Added a new unit test for this as well.

islemaster commented on apps/src/applab/CrosshairOverlay.js:
```diff
@@ -125,6 +154,17 @@ CrosshairOverlay.prototype.create_ = function () {
+  this.elementIdBubble_.setAttribute('width', TEXT_RECT_WIDTH);
+  this.elementIdBubble_.setAttribute('height', TEXT_RECT_HEIGHT);
+  this.elementIdBubble_.setAttribute('rx', TEXT_RECT_RADIUS);
+  this.elementIdBubble_.setAttribute('ry', TEXT_RECT_RADIUS);
```
This code is duplicated from above.   Please extract a small helper, something like `createTextBubbleElement_`.

kozzi replied:
Fixed.

islemaster commented on apps/src/applab/CrosshairOverlay.js:
```diff
@@ -95,6 +101,29 @@ CrosshairOverlay.prototype.render = function (intoElement, nextProps) {
   this.text_.textContent = this.getCoordinateText();
+
+  // If user is hovering over a control, show the element's id as a second tooltip
+  if (this.props_.mouseoverApplabControlId) {
```
This doesn't quite duplicate the above, and in fact depends on it, so it might be fine for now.  We could try to extract the text bubble stuff into a sub-component now, or we could wait until we have three of them.  What do you think?

kozzi replied:
Yeah, I left this one alone.

islemaster commented:
You've probably noticed that I didn't write good tests for this feature the first time around :fearful: and I'm not sure I have good advice for unit-testing SVG rendering on a feature that's heavily dependent on mouse movement.

Actually, the best way to test this feature might be with a [new eyes test](https://github.com/code-dot-org/code-dot-org/blob/staging/docs/testing-with-applitools-eyes.md).  You could reuse some of the steps from [this applab share page test](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/ui/features/applab/sharedApps.feature#L83) to switch between design and code mode, drag elements into the app, etc. and then use Selenium to move the mouse over the element and do an eyes check.

kozzi commented:
Added some new mocha and eyes test. For the eyes test, I ran it locally, so will likely require new baselines once running on test server. 

I also ran eyes test that use `I drag the grippy by...` step since I extracted some code from that. They continued to work in my tests.

I did as much testing as I could. Are there any other scripts/tests I should run to make sure everything still works?

islemaster commented on apps/src/applab/CrosshairOverlay.js:
```diff
@@ -116,15 +145,11 @@ CrosshairOverlay.prototype.create_ = function () {
-  this.bubble_.setAttribute('ry', TEXT_RECT_RADIUS);
-  this.ownElement_.appendChild(this.bubble_);
+  // Create the cooordinates tooltip
+  [this.bubble_, this.text_] = this.createTextBubbleElement_();
```
[Destructuring assignment is part of ES2015](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Specifications) which we're not _technically_ supporting in our code yet (though we're all anxious to get to it).

kozzi replied:
Fixed!

islemaster commented:
This is so awesome!  Nice work @kozzi .  LGTM besides the destructing assignment note.
